### PR TITLE
[Unity][Frontend] Translate StableHLO to Relax

### PR DESCRIPTION
**Pull request by yongwww**
Add a converter to convert the `StableHLO` to Relax, the source `StableHLO` IR  could be lowered from `JAX`, or `TensorFlow`.

-  It depends on `StableHLO` Python Bindings maintained in `jaxlib`
-  It only supports one-way conversion, and no MLIR Relax is introduced
-  Use JAX to generate reference inference result for accuracy verification, and an e2e (`JAX ResNet -> StableHLO -> Relax`) works

### CODE REVIEW

File: python/tvm/relax/frontend/stablehlo/stablehlo_translator.py [hunk 1]
```diff
@@ -0,0 +1,417 @@
...
+import numpy as np
+
+from jaxlib import mlir
+from jaxlib.mlir.dialects import stablehlo
+
+import tvm
+from tvm import relax, tir
+
+
+class StableHLOImporter:
```
Comment (Hzfengsy) [inline python/tvm/relax/frontend/stablehlo/stablehlo_translator.py:31]: As we may have more importers, should we create a base class to reuse some utils? cc @spectrometerHBH @MasterJH5574

Comment (yongwww) [reply python/tvm/relax/frontend/stablehlo/stablehlo_translator.py:28]: That's a good idea. We can certainly proceed with it. For now in this pr, we may consider this as a todo, as the currently available shared helper functions are somewhat limited.

**Comment by yongwww** [comment]
@tvm-bot rerun

File: python/tvm/relax/frontend/stablehlo/stablehlo_translator.py [hunk 2]
```diff
@@ -0,0 +1,431 @@
...
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+# pylint: disable=invalid-name, inconsistent-return-statements, unidiomatic-typecheck
+# pylint: disable=import-outside-toplevel, unused-argument, no-member
+"""StableHLO frontend of Relax."""
```
Comment (psrivas2) [inline python/tvm/relax/frontend/stablehlo/stablehlo_translator.py:20]: should we make this file pylint compliant?

File: python/tvm/relax/frontend/stablehlo/stablehlo_translator.py [hunk 3]
```diff
@@ -0,0 +1,431 @@
...
+            "stablehlo.sine": self._sin,
+            "chlo.sinh": self._sinh,
+            "stablehlo.sqrt": self._sqrt,
+            "stablehlo.subtract": self._subtract,
+            "func.return": self._return,
+            "stablehlo.return": self._return,
+        }
+
+    def from_stablehlo(self, model, input_info: List[Tuple[Tuple[int], str]]) -> tvm.IRModule:
+        """Convert a StableHLO Module to a Relax program."""
```
Comment (psrivas2) [inline python/tvm/relax/frontend/stablehlo/stablehlo_translator.py:355]: adding doc for parameters would be useful

**Review by psrivas2 — COMMENTED** [review]
Thanks @yongwww! This is a nice first step towards importing jax/tf models.
Looks in good shape functionality wise. Would be nice to improve documentation and clean up imports.

File: python/tvm/relax/frontend/stablehlo/stablehlo_translator.py [hunk 4]
```diff
@@ -0,0 +1,431 @@
...
+                        # TODO (yongwww): handle multiple outputs
+                        output = self.block_builder.emit_output(self._nodes[operation])
+                        break
+
+                    if isinstance(operation, mlir.ir.OpView):
+                        op_name = operation.operation.name
+                        assert op_name in self.convert_map, f"Unsupported operation {op_name}"
+                        self._nodes[operation] = self.convert_map[op_name](operation)
+                    else:
+                        raise ValueError(f"Unsupported op {operation}")
```
[the remark this answers is not in the document]
Comment (yongwww) [reply python/tvm/relax/frontend/stablehlo/stablehlo_translator.py:408]: I added this as todo in the translator, currently I don't get a test case with impurity, and need to figure out a method to tell if the op is impure in python binding.

File: python/tvm/relax/frontend/stablehlo/stablehlo_translator.py (hunk 2 above)
Comment (yongwww) [reply python/tvm/relax/frontend/stablehlo/stablehlo_translator.py]: Good idea, have removed most of them now. Still have two: import-outside-toplevel, unused-argument. The first one is due to we import stablehlo and mlir inside the function, the second one is the input_info has not been used at this moment, will use it to feed input shapes (static/symbolic) for improvement.

File: python/tvm/relax/frontend/stablehlo/stablehlo_translator.py [hunk 5]
```diff
@@ -0,0 +1,431 @@
...
+        elif input_type == "ui32":
+            return "uint32"
+        elif input_type == "ui64":
+            return "uint64"
+        else:
+            raise NotImplementedError(f"input_type {input_type} is not handled yet")
+
+    def _attr2value(self, node) -> Union[Any, List[Any]]:
+        from jaxlib import mlir
+        import numpy as np
```
[the remark this answers is not in the document]
Comment (yongwww) [reply python/tvm/relax/frontend/stablehlo/stablehlo_translator.py:77]: When we try to import relax, the frontend will be imported as part of relax, import will failed if jaxlib is not installed. Making the imports inside function could help avoid this inconvenience (similar the handling as other translators)

**Comment by yongwww** [comment]
CI is green, pls take a look. cc: @psrivas2  @Hzfengsy  @tqchen

**Review by tqchen — APPROVED** [review]

**Comment by junrushao** [comment]
Note on unittests: in the future, we want to migrate `check_correctness`-based numerical tests to IR structural equality test, which is much more accurate and to-the-point

**Pull request merged**